---
title: "RNAseq Simulation"
author: "Francesco Lescai"
date: "23/08/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd("~/PROJECTS_WORK/RNAseq_simulation")
library(polyester)
library(Biostrings)
```



```{r}
fasta = readDNAStringSet("~/DATABASES/human/Homo_sapiens.GRCh38.cdna.all.fa")
readspertx = round(30 * width(fasta) / 100)
```



```{r}
genes<- names(fasta)
immunoglobulin <- sample(which(grepl("immunoglobulin", genes)), 200)
potassium <- sample(which(grepl("potassium", genes)), 150)
```


```{r}
fold_changes <- matrix(
  rep(1,2*length(genes)),
  nrow = length(genes)
)
fold_changes[immunoglobulin,1] <- sample(2:4,length(immunoglobulin), replace = T)
fold_changes[potassium,2] <- sample(2:4,length(potassium), replace = T)
```


```{r}
simulate_experiment(fasta = "~/DATABASES/human/Homo_sapiens.GRCh38.cdna.all.fa",
                    outdir = "reads",
                    num_reps = c(20,20),
                    reads_per_transcript=readspertx,
                    fold_changes=fold_changes,
                    size = 0.2,
                    error_model = "illumina5",
                    gzip=TRUE
                    ### if doMC is installed here you can add
                    ### cores=16 depending on the available cores
                    ### but devtools::install_github('kcha/polyester')
                    ### should be run first
                    )
```



