
---
class: center, middle, inverse, title-slide
title: "Internship Summary Presentation"
subtitle: "Testing RNA-Seq Pipelines and Simulators"
author: "Jonathan Brenton"
institute: "National Institute for Biological Standards and Control (NIBSC)"
date: "2019/12/12 (updated: `r Sys.Date()`)"
output:
  xaringan::moon_reader:
    css: nibsc_report.css
    lib_dir: libs
    nature: 
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: true

---
```{r loading workspace, include=FALSE, eval=FALSE}
load("flux_salmon_quants_03.12.19/.RData")
```

class: bottom, middle

background-image: url(https://www.genengnews.com/wp-content/uploads/2018/10/Getty_537618080_ktsimage_DNATestSangerSeq.jpg)
background-size: 400px
background-position: 50% 20%

<P style="font-size: 34pt; color: red">NGS

---
Next Generation Sequencing is ever-present in current biological research

-unbiased approach that covers the whole genome

-exciting new discoveries and applications compared to old methods


---
class: center, middle, inverse

But how accurate and reliable are these tests?

---

We simulated data from two separate tools



---

class: center, middle, inverse

show graphs:

Which looks better to you?


---
class: center, middle, inverse

Create Reads:

Polyester vs Flux-Simulator
-pro vs cons
-Polyester can create differential expression matrix but not upstream parameters
-Flux can change various sequencing parameters but has not automated differential expression matrix
How both are done? Matrix

---
class: center, middle, inverse

Pipeline
-Create reads
-Shuffle reads
-Align and quantify reads
-Differential expression


---
class: center, middle, inverse
Shuffling reads
Why did?
How it's done/visual
Also could use seqkit but need higher memory but worth testing


---
class: center, middle, inverse
Salmon
How it's done/why best?
Visual
Hash table

---
class: center, middle, inverse
Tximport
-tx2gene
-does this filter anything?

---
class: center, middle, inverse

DEQeq 

---
class: center, middle, inverse
#Initial results:
##30x coverage for polyester
```{r fold change, include=FALSE, eval=FALSE}
head(fold_changes)
```

Need lots of memory for shuffling
Wrong genome/too many reads in wrong place
Used 

---


class: center, middle, inverse

Then tried again with different gencode but similar problem



---

class: center, middle, inverse
Then tried a shortened version...
Still same problem-graph


---

class: center, middle, inverse
Tried flux simulator manually
-visual of mannually attaching the reads


---

class: center, middle, inverse
Differential expression:



---
class: center, middle, inverse
Conclusions

---
One main problem:
-Aligment and differential analysis software tools are not often compared 
-When they have been compared against each other in the context of real data
-Therefore, it is not clear what the actual outcome is!


---




