---
title: "Polyester Simulation Code"
author: "Jonathan Brenton"
date: "27/12/2019"
output:
  html_document:
    theme: readable
    highlight: tango
    toc: true
    toc_float: true
    css: C:/Users/JBrenton/Polyester_analysis_30_09_19/nibsc.report.css
editor_options:
  chunk_output_type: console
---

# Setup

```{r setup, include=TRUE, eval=FALSE}
setwd("/usr/share/sequencing/internships/JonathanBrenton/polyester_simulation/simulate/polyester_tests_13.10.19_excit+inhib_genes")
library(doMC)
library(polyester)
library(Biostrings)
library(stringr)
library(dplyr)

### read the fasta file and create a coverage of 30 FPKM
### i.e. proportional to the transcripts length

writeLines("####### reading transcriptome FASTA")

gencode_fasta_file="/usr/share/sequencing/references/to_be_deleted/rnaseq/Homo_sapiens/ENSEMBL/gencode.v32.transcripts.fa.gz"
gencode_fasta = readDNAStringSet(gencode_fasta_file)

readspertx = round(30 * width(gencode_fasta) / 100)
```

This section of the code sets up the polyester simulation by reading in the fasta transcriptome, here we use the Gencode fasta to limit the non-verified genes in the fasta and have better matches between the Hugo symbols, reads and entrezids. The readspertx variable is used to provide a coverage proportional to gene length, so in this example a coverage of 30FPKM would be. 

### Extracting Fasta Header Information

```{r gene info splitting, include=TRUE, eval=FALSE}
###get the gene information separated

genes_gencode<- names(gencode_fasta)

parseGencode<-function(genestring){

  identifier<-sub("^(ENST.+)\\|.+\\|.+\\|.+\\|.+\\|(.+)\\|.+\\|(.+)\\|$", "\\1", genestring)
  geneName<-sub("^(ENST.+)\\|.+\\|.+\\|.+\\|.+\\|(.+)\\|.+\\|(.+)\\|$", "\\2", genestring)
  description<-sub("^(ENST.+)\\|.+\\|.+\\|.+\\|.+\\|(.+)\\|.+\\|(.+)\\|$", "\\3", genestring)

  return(c(identifier, geneName, description))
}
######apply the function to the gencode information to extract the terms above

extracted_gencode<-do.call("rbind", lapply(genes_gencode,parseGencode))

colnames(extracted_gencode)<-c("Transcript_ID", "Gene_Name", "Description")

extracted_gencode<-as.data.frame(extracted_gencode)
```

The above section uses a text wrangling function to extract the sections from the fasta header we are interested in and create a reference data frame to use for position and transcript name information.

### Getting Genenames for upregulated genes

```{r gene selection, include=TRUE, eval=FALSE}
read.csv(file="inhibitory_updated_list.csv", header=TRUE, skip=1)-> inhibitory_list

inhibitory_list<-inhibitory_list %>% filter(!str_detect(Match.type, 'Unmatched'))

read.csv(file="excitatory_updated_list.csv", header=TRUE, skip=1)-> excitatory_list

excitatory_list<-excitatory_list %>% filter(!str_detect(Match.type,"Unmatched"))

### C6ORF165 was duplicated so:
excitatory_list<-excitatory_list[!duplicated(excitatory_list$Approved.symbol),]

####mismatch between ensembl and HUGO-KIF1BP vs KIFBP, because using gencode had to change

levels(inhibitory_list$Approved.symbol) <- c(levels(inhibitory_list$Approved.symbol), 'KIF1BP')
inhibitory_list$Approved.symbol[inhibitory_list$Approved.symbol == 'KIFBP'] <- 'KIF1BP'

##so to remove overlapping :
overlap<-intersect(excitatory_list$Approved.symbol, inhibitory_list$Approved.symbol)
excitatory_list<-excitatory_list[excitatory_list$Approved.symbol %in% setdiff(excitatory_list$Approved.symbol,overlap),]
inhibitory_list<-inhibitory_list[inhibitory_list$Approved.symbol %in% setdiff(inhibitory_list$Approved.symbol,overlap),]
```

For this simulation we picked two sets of neuronal genes, those specific to excitatory and inhibitory neurons. The excitatory genes will be upregulated in one group and the inhibitory in another. This section of the code takes a list of genes from a file which will later be checked to see if they are found to be significantly upregulated. These genes should then be found to be involved in certain gene pathways.

### Removing version info and double checking presence of upregulated genes in Fasta file

```{r making names, include=TRUE, eval=FALSE}
extracted_gencode$Transcript_ID<-sub("\\..+", "", extracted_gencode$Transcript_ID)

excitatory_names_and_gene_matches<-extracted_gencode %>% filter(Gene_Name %in% excitatory_list$Approved.symbol)

inhibitory_names_and_gene_matches<-extracted_gencode %>% filter(Gene_Name %in% inhibitory_list$Approved.symbol)
```

The version information of the transcripts is removed to make life easier in later analysis. The names between the extracted gencode information from the fasta and the names of gene names in the two lists we selected are cross-referenced to make sure they are present in the fasta file we will use to generate the reads using polyester. 

### Creating a reference csv file of gene names and writing positions of upregulated genes into separate objects

```{r tables and positions in gencode, include=TRUE, eval=FALSE}
write.table(excitatory_names_and_gene_matches, "Excitatory_genes_transcripts_upregulated.csv", col.names =TRUE, sep=" ", row.names=FALSE)
write.table(inhibitory_names_and_gene_matches, "Inhibitory_genes_transcripts_upregulated.csv", col.names =TRUE, sep=" ", row.names=FALSE)

excit_positions<-which(extracted_gencode$Transcript_ID %in% excitatory_names_and_gene_matches$Transcript_ID)
inhib_positions<-which(extracted_gencode$Transcript_ID %in% inhibitory_names_and_gene_matches$Transcript_ID)
```

Here, we write some tables of the gene names to be used later in the analysis and to reference later in case of problems. The positions/rownumbers of these genes within the fasta file are then found.

# Polyester Simulation code

### Setting up fold change matrix

```{r fold changes, include=TRUE, eval=FALSE}
writeLines("##### preparing fold changes matrix")
## I then create a baseline fold matrix with 1
fold_changes <- matrix(
  rep(1,2*length(genes_gencode)),
  nrow = length(genes_gencode)
)


fold_changes[excit_positions,1] <- sample(2:4,length(excit_positions), replace = T)
fold_changes[inhib_positions,2]  <- sample(2:4,length(inhib_positions), replace = T)
## identical(which(fold_changes[,2]>1),inhib_positions)
##[1] TRUE
```

Here the fold change matrix is created, which will be used by the polyester package to select which genes are upregulated. It has the same length as the fasta file and two columns for the two groups we will be creating. 

### Running the Simulation

```{r simulate experiment, include=TRUE, eval=FALSE}
simulate_experiment(fasta = gencode_fasta_file,
                    outdir = "excit.inhib.polyester.test.size-0.3",
                    num_reps = c(20,20),
                    reads_per_transcript=readspertx,
                    fold_changes=fold_changes,
                    size = 0.3,
                    error_model = "illumina5",
                    gzip=TRUE,
                    ### if doMC is installed here you can add
                    cores=32
                    ### but devtools::install_github('kcha/polyester')
                    ### should be run first
                    )
```

Here we use the simulate experiment function to create reads across the fasta file/transcriptome for 20 samples in one group and 20 in another. The first group will have excitatory neuronal genes upregulated and the second group will have inhibitory neuronal genes upregulated. We have been using a forked version of polyester which can be run with the doMC package, which allows the use of multi-threading. Here we specifiy the fasta file used, the output directory, the sequencing depth per transcript, the fold_change matrix used to specify upregulated genes, the size parameter (see polyester manual; specifies a mean-variance relationship for experiment-size 0.3 is default), the error model (inbuilt illumina the one was used, since illumina is used in the lab), whether the output fasta files should be compressed, and the amount of cores to use. This code a lot of memory so a lot of cores were used. 

