---
title: "Salmon_script_markdown"
author: "Jonathan Brenton"
date: "27/12/2019"
output:
  html_document:
    theme: readable
    highlight: tango
    toc: true
    toc_float: true
    css: C:/Users/JBrenton/Polyester_analysis_30_09_19/nibsc.report.css
editor_options:
  chunk_output_type: console
---
# Creating a transcript index for salmon

```{bash creating index, include=TRUE, eval=FALSE}
salmon index -t gencode.v32.transcripts.fa.gz -i gencode.v32.transcripts.index
```

This creates an index of the transriptome for salmon to look up reads against. The input is the *.fa.gz* file and the output index file can be named as anything, in this case gencode.v32.transcripts.index. 

# Initial Bash Script setup for slurm scheduler

```{bash setup, include=TRUE, eval=FALSE}
#!/bin/bash -l
#Force bash as the executing shell.
#SBATCH -N 1
#SBATCH -c 4
#SBATCH --mem=24g
#SBATCH --job-name=Salmon

echo "### START - the job is starting at"
date
starttime=`date +"%s"`
echo

echo "the job is running on the node $SLURM_NODELIST"
echo "job number $SLURM_JOB_ID"
echo "STAT:jobName:slurm_Salmon_$SLURM_JOB_ID\.out"
echo "STAT:exechosts:$SLURM_NODELIST"
echo
```

The first section of the code gives the node, cores and memory parameters for the slurm scheduler to use, along with the name of the jobname which will appear in the queue. The next section of the code provides the start time and the job ID and names for each job that will be output into the standard output.

# Input to salmon program

```{bash salmon input, include=TRUE, eval=FALSE}
samplequery=$1
WORKDIR=$2
sample=$3
transcriptsIndex=$4
LIBTYPE=$5
```

These variables are passed to the script/job in the command line. The first variable is the two fasta or fastq files separated by a comma. *E.g Sample_01_1.gz,Sample_01_2.gz*. The second variable is the directory comtaining the fasta files. The third is the basename of the fasta file which should be used to make a folder for the output. *E.g Sample_01*. The fourth is the index the transcript index created earlier. Finally libtype is defined by the type of sequencing performed, more info is here: https://salmon.readthedocs.io/en/latest/library_type.html#fraglibtype

```{bash string split, include=TRUE, eval=FALSE}
STR_ARRAY=(`echo $samplequery | tr "," "\n"`)
query1=${STR_ARRAY[0]}
query2=${STR_ARRAY[1]}
inputFiles="$query1 $query2"

echo "the input will be $inputFiles"
```

This part of the script takes the sample query files and splits them into separate files. It swaps the comma inbetween Sample_01_1.gz,Sample_01_2.gz and splits them into names on separate lines.

```{bash folders, include=TRUE, eval=FALSE}
parentFolder=${WORKDIR%/sequences}

echo "assuming the parent folder is $parentFolder"
mkdir -p $parentFolder/alignments
cd $parentFolder/alignments
echo "now emitting alignments into `pwd`"
```

This takes the working directory and makes a new folder for the salmon alignments.

```{bash salmon, include=TRUE, eval=FALSE}
alignstart=`date +"%s"`
echo "############## running the Salmon quasi-alignment of the two pairs"


### Salmon base code here

salmon quant \
-p 4 \
-i $transcriptsIndex \
-l $LIBTYPE \
-1 $WORKDIR/$query1 \
-2 $WORKDIR/$query2 \
-o ${sample}_salmon_quant
```

Here, the p option specifies the number of cores, while the others specify the index, the library type, the two input files and an output folder.

```{bash echo, include=TRUE, eval=FALSE}
date
alignend=`date +"%s"`
alignrun=$((alignend - alignstart))
echo "STAT:ALIGNSTART:$alignstart"
echo "STAT:ALIGNEND:$alignend"
echo "STAT:ALIGNRUN:$alignrun"

echo "####END job finished"
endtime=`date +"%s"`
duration=$((endtime - starttime))
echo "STAT:startTime:$starttime"
echo "STAT:doneTime:$endtime"
echo "STAT:runtime:$duration"
```

The last part of the script provides the start time and end time for the script, for troubleshooting. 

# Command line entry example
```{bash command line entry, include=TRUE, eval=FALSE}
for num in {01..40}
do
sbatch -o "`pwd`/logs/salmon_job_%j.out"
slurm_salmon.job sample_${num}_1_shuffled.fasta.gz,sample_${num}_2_shuffled.fasta.gz 
"/home/AD/jbrenton/fastashuffle/output" sample_${num} 
`pwd`/Homo_sapiens.GRCh38.cdna.all_index IU
done
```

This is an example of the command line entry used to parallelize the job/script. The options after the slurm_salmon.job are the options for salmon. 
